### ISSUE TITLE
Playlist with videojs-ima stops playing prerolls after scrubbing any item

### ISSUE THREAD

**Issue by jg-proboxtv** [googleads/videojs-ima#1161]
In the code below, I have created a video player component within a nextJS 15 app router application.  The code works quite well.  When the page is loaded, it automatically plays the first video in the playlist.  It will then correctly move through subsequent playlist items (I haven't yet bothered stopping it from moving past the end of the playlist - that is not the issue here).  In addition, a preroll ad properly plays before each video.  The preroll plays before the autoplayed initial video, and the preroll plays before each subsequent video.  All of this works as desired.

Unfortunately, there is a bug.  If I play a video (call it video1), but I scrub forward in video 1 almost to the end (basically 5 or so seconds before the end, but even 30 seconds before the end will cause same issue), and then let video 1 complete ... video 1 ends (all good) ... a preroll plays (good) ... then the next video plays (call it video2, and all is still good).  But, then, if I click on to another video in the playilst (call it video3), I do not get a preroll before video 3.  In other words, once I scrub forward in a video and then let that video complete, something gets messed up.

Please bear in mind, ads and videos all work great as I click around in the playlist.  The issue only occurs if I scrub within a video, let that video finish, wait for the preroll of the next video to finish, and then, after that next video starts, finally, I click on another video in the playlist.  It is the act of scrubbing a video and then letting it finish naturally that somehow messes things up.

I get the same errors in the console every time I click that last video (what I called video3 in my example).  The errors are:
1. AdsLoader error: AdError 303: No Ads VAST response after one or more Wrappers
2. VIDEOJS: ERROR: An error with Ads occured. Type: ads-preroll-error.
3. GET http://localhost:3001/__nextjs_original-stack-frame?isServer=false&isEdgeServer=false&isAppDirectory=true&errorMessage=Error%3A+VIDEOJS%3A+%22ERROR%3A%22+%22An+error+with+Ads+occured.+Type%3A+ads-preroll-error.%22&file=chrome-extension%3A%2F%2Ffmkadmapgofadopljbjfkapdkoienihi%2Fbuild%2Freact_devtools_backend_compact.js&methodName=console.overrideMethod+%5Bas+error%5D&arguments=&lineNumber=13851&column=11 400 (Bad Request)

Those same three errors occur every time the preroll fails to play.  Interestingly, if I just let the video (say, video3) continue playing until the end, then a preroll will play properly before the next video.  I can also click any other playlist item, and again, the preroll will play properly.  In other words, as long as I don't scrub again, everything goes back to normal.  The moment I scrub again, I get the same behavior and the same three errors.

I have tried Open AI (even the new o1 pro model).  It does not help at all.  In fact, some of its suggestions seem just plain wrong.  I have read pretty much everything I could find online.  My last hope will just be to read the videojs-ima library source code.  However, I am not convinced even that will lead me to the solution.  I hope an expert here might have experienced this before or might just immediately see my mistake.  I feel like I might just be failing to properly handle the pause/play events that happen on scrubbing.  Maybe, I am not handling the promise returned from the new play event that is initiated after the scrub is done.

Any help is truly appreciated.  Thank you.

Quick note: I have left a fair amount of commented code in place below.  I did this to show a few things I have tried - none of which helped solve the problem.

VideoPlayer2.tsx
-----------------
```
"use client"

import React from "react";
import {useRef, useEffect, useState} from "react";
import Script from "next/script";

import "/public/dist/videojs.css";
import "/public/dist/videojs.ima.css";
import "/public/dist/styles.css";
import videojs from "video.js";
import "/public/dist/playlist";
import "/public/dist/videojs.thumbnails";
import "/public/dist/videojs.ads";
import "/public/dist/videojs.ima";

import {myPlaylist} from "/public/dist/medialist.js";
import Player from "video.js/dist/types/player";

class VideoPlayer {

  constructor(videoEl, elId) {

    this.player = videojs(videoEl, {
      fluid: true,
      //poster: "//cdnzone.nuevodevel.com/images/coffee.jpg",
      autoplay: true,
      muted: true,
      controls: true
    });

    const imatag =
        "https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&iu=/124319096/external/single_ad_samples&ciu_szs=300x250&impl=s&gdfp_req=1&env=vp&output=vast&unviewed_position_start=1&cust_params=deployment%3Ddevsite%26sample_ct%3Dskippablelinear&correlator=";

    var options = {
      id: elId,
      adTagUrl: imatag
    };

    this.player.on("ended", this.onEnd);
    this.player.on("play", this.onPlay);
    this.player.on("pause", this.onPause);
    this.player.on('loadedmetadata', this.onLoadedMetadata);

    this.player.ima(options);
    this.player.ima.initializeAdDisplayContainer();
    this.nextIndex = 0;
    this.play(true)
  }

  onLoadedMetadata = () => {

  }

  play = (constructed = false) => {
    this.player.trigger('contentchanged'); // OAI
    let videoSource = myPlaylist[this.nextIndex].sources[0].src;
    this.player.ima.setContentWithAdTag(videoSource, null);
    this.player.poster(myPlaylist[this.nextIndex].poster);
    
    if (!constructed) {
      //this.player.ima.initializeAdDisplayContainer(); // Open AI suggestion, makes no difference
      this.player.ima.requestAds();
    }

    var playPromise = this.player.play();
    if (playPromise !== undefined) {
      playPromise.then(_ => { })
      .catch(error => { });
    }
  }

  onEnd = (event) => {
    /**Open AI suggestion - crashes */
    // if (this.player.ima && this.player.ima.adsLoader) {
    //   this.player.ima.adsLoader.contentComplete();
    // }
    /**ai */

    this.nextIndex++;
    this.play();
  }

  onPlay = (event) => {
    console.log('this.onplay: ', event)
  }

  onPause = (event) => {
    console.log('this.onpause: ', event)
  }

  onPlaylistItemClick = (event, index) => {
    /**Open AI suggestion - crashes */
    // this.player.trigger('contentended');
    // if (this.player.ima && this.player.ima.adsLoader) {
    //   this.player.ima.adsLoader.contentComplete();
    // }
    /**ai */

    let lastChar = event.target.id.charAt(event.target.id.length - 1);
    let intValue = parseInt(lastChar);
    this.nextIndex = intValue;
    this.play();
    // if (!linearAdPlaying) {
    //   if (!initialized) {
    //     console.log('Was not initialized')
    //     // Handles the case where the user loads the page, clicks a playlist item
    //     // immediately, and never clicks the play button on the player.
    //     init();
    //   }
    //}
  }

  dispose = () => {
    console.log('Disposing of player')
    this.player.dispose()
  }
}

export default function VideoPlayer2(params) {
  const [player, setPlayer] = useState(null)

  const videoRef = useRef(null);
  const playlistRef = useRef(null);
  const consoleRef = useRef(null);

 
  useEffect(() => {

    if (videoRef.current) {
      const player = new VideoPlayer(videoRef.current, params.id)
      setPlayer(player);

      /*
      // Remove controls from the player on iPad to stop native controls from stealing
      // our click
      var contentPlayer =  document.getElementById('content_video_html5_api');
      if ((navigator.userAgent.match(/iPad/i) ||
            navigator.userAgent.match(/Android/i)) &&
          contentPlayer.hasAttribute('controls')) {
        contentPlayer.removeAttribute('controls');
      }
      
      // Initialize the ad container when the video player is clicked, but only the
      // first time it's clicked.
      var initAdDisplayContainer = function() {
        player.ima.initializeAdDisplayContainer();
        wrapperDiv.removeEventListener(startEvent, initAdDisplayContainer);
      }
      
      var startEvent = 'click';
      if (navigator.userAgent.match(/iPhone/i) ||
          navigator.userAgent.match(/iPad/i) ||
          navigator.userAgent.match(/Android/i)) {
        startEvent = 'touchend';
      }
      
      var wrapperDiv = document.getElementById('content_video');
      wrapperDiv.addEventListener(startEvent, initAdDisplayContainer);
      */

      // return () => {
      //   player.dispose()
      // };
    }

  }, []);

    return (
        <>
            {/* <Script src="https://www.googletagservices.com/tag/js/gpt.js" strategy="beforeInteractive"></Script> */}
            <Script src="https://imasdk.googleapis.com/js/sdkloader/ima3.js" strategy="beforeInteractive"></Script>

            <video id={params.id} className="video-js vjs-default-skin" 
                ref={videoRef} controls preload="auto" width="100" height="100">
            </video>

            { player &&
                <div id="playlist" ref={playlistRef}>
                    <ul>
                        {
                        myPlaylist.map(
                            (item, idx) => {
                            return <li id={`playlist-item-${idx}`} key={idx} onClick={(event) => player.onPlaylistItemClick(event, idx)}>{item.title}</li>
                            }
                        )
                        }
                    </ul>
                </div>
            }

        </>
    );
}

```
page.tsx
---------
```
"use client"

import VideoPlayer2 from "@/components/videojs/VideoPlayer2";
import React from "react";
import {useRef, useEffect, useState} from "react";


export default function IndexPage() {
    return (
        <>
            <VideoPlayer2 id="content_video" />
            {/* <VideoPlayer2 id="content_video2" /> */}
        </>
    );
}

```

**Comment by Kiro705** [comment]
Hello @jg-proboxtv ,

I wanted to check if you are seeing this error in production. The AdError 303 could indicate that you are making too many calls to the test ad tag, and so an empty VAST tag is being returned. A quick test could be to use a different [IMA sample tag](https://developers.google.com/interactive-media-ads/docs/sdks/html5/client-side/tags) for each video, to see if the error still happens.

This issue is specific to the IMA test tags, and I would not expect it to happen for production ads.

Thank you,
Jackson
IMA SDK team

**Comment by jg-proboxtv** [comment]
Jackson,

Thank you for the reply.  I actually managed to fix the issue yesterday, but hadn't yet found a moment to post the solution here.  I still have another issue - if I click on a playlist item while an advertisement is running, I get into a funny state where the scrub bar for the ad and the scrub bar for the video both show, the video itself shows, but doesn't play - just kind of hangs.  However, I intend to try to solve that issue using the same approach I used to solve this one.

Essentially, I modified my onEnd and onPlaylistItemClick handlers as such:
```
onEnd = (event) => {
    // I added this code to both onEnd and onPlaylistItemClick
    if (this.player.ima && this.player.ima.controller.sdkImpl.adsLoader) {
      console.log('onEnd calling contentComplete')
      this.player.ima.controller.sdkImpl.adsLoader.contentComplete();
    }
   // End of new code

    this.nextIndex++;
    this.play();
  }

```
It felt a bit dirty to dig into the ima controller and then even further to get the adsLoader, but I guess this is what I need to do.  If there is a better way to access the underlying adsLoader, or another way to do what I have done, I'd love to know.  From this point, I will try to fix any new issues by just directly accessing the underlying ima objects.  Thank you again for your reply, and please do let me know if you see a better solution.

**Comment by Kiro705** [comment]
Hello @jg-proboxtv ,

This video.js-IMA plugin is meant to support the main use-cases of the SDK, but because of that may not support all custom implementations. Depending on the amount of customization you need for your app, it is also possible to use a video.js player with the IMA SDK, without using this plugin.

Right now, I'm not sure the best way to achieve switching to a new content source while an ad is running, within the same video.js player. Work-arounds may be to prevent switching content during an ad, or loading a new page when switching to a  new content source.

Thank you,
Jackson
IMA SDK team

**Comment by DharmikUnagar** [comment]
1. Why contentComplete() is Essential
The videojs-ima plugin (and indeed the IMA SDK) needs to be told when your main content is finished playing. If you skip that notification (i.e. don’t call contentComplete()), the SDK’s internal ad state machine will not clean itself up, and future ads can get into a weird or broken state.

Whether you call

js
Copy code
player.ima.adsLoader.contentComplete();
or use the more “internal” route

js
Copy code
player.ima.controller.sdkImpl.adsLoader.contentComplete();
the important thing is simply to invoke it at the correct time—i.e., once your content is really done.

Why some folks end up calling controller.sdkImpl.adsLoader:
In older versions of videojs-ima, the plugin didn’t expose adsLoader publicly. So people resorted to referencing the internal controller object. Fortunately, in more recent versions you often can (and should) call the public player.ima.adsLoader.

2. Prefer the Public API (If Available)
Before dropping down to controller.sdkImpl, try:

js
Copy code
if (this.player.ima && this.player.ima.adsLoader) {
  this.player.ima.adsLoader.contentComplete();
}
A lot of folks find that it “crashes” or is undefined if the plugin version is older or if the plugin hasn’t yet fully loaded its internal objects. If that’s happening, you can wait until the ad plugin has fully initialized, or ensure you’re only calling it once you know the plugin is ready. That might involve a small tweak to your plugin lifecycle (for example, waiting for an event like adsready).

If that never works in your environment, then yes—controller.sdkImpl.adsLoader is the fallback. But keep in mind, it’s not guaranteed to be stable across plugin upgrades.

3. Use the contentended Event (Where Possible)
The [videojs-ima documentation](https://github.com/googleads/videojs-ima#quick-start) suggests that if your player fires a contentended event on its own (many do), the plugin will automatically call contentComplete() behind the scenes. In that scenario, you can do something like:

js
Copy code
onEnd = () => {
  // Let video.js know content is done. The plugin will see this and handle the ad teardown.
  this.player.trigger('contentended');

  // Move on to next item:
  this.nextIndex++;
  this.play();
};
Sometimes, though, video.js players or custom-coded logic do not automatically fire contentended. If that’s your situation, manually triggering the event (as above) is often the simplest way to achieve the same result.

4. Handling the “Click During Ad” Issue
You mentioned you still see odd behavior if a user clicks on a playlist item while an ad is still running. Typically, you want to ensure that:

If the ad is currently playing, you either (a) let the ad finish, or (b) gracefully stop the current ad playback. Stopping an ad break mid-stream often means you should call adsManager.stop() or something similar via the plugin.
If you do plan to skip the ad (by letting the user click a new video in the playlist), you’ll want to call contentComplete() for the old item so the plugin cleans up. Then, proceed with the new content (which triggers contentchanged, etc.).
Pseudocode-ish example:

js
Copy code
onPlaylistItemClick = (event, index) => {
  // If an ad is running, gracefully stop it.
  if (this.player.ima && this.player.ima.adsActive) {
    try {
      // The plugin has some methods for this. 
      // For some older versions, you'll do something like:
      // this.player.ima.controller.adsManager.stop();
      // or a combination of methods. 
      this.player.ima.adsManager.stop();
    } catch (err) {
      console.warn('Could not stop ads', err);
    }
  }

  // Let IMA know content was completed, so it resets properly.
  if (this.player.ima && this.player.ima.adsLoader) {
    this.player.ima.adsLoader.contentComplete();
  }

  // Now we actually move to the new video...
  this.nextIndex = index;
  this.play();
};
Which approach you take (stopping vs. waiting) depends on the user experience you want. But the main idea is that if you want to forcibly skip or end the current content’s ad, you should also tell IMA that that content is effectively finished so the plugin state resets. This prevents the “double-scrub bar” scenario where the ad’s progress bar might remain stuck.

5. Summary / Best Practices
Always call contentComplete() on your main content finishing (and also if you’re forcibly moving on).
Use the plugin’s public adsLoader if possible, rather than controller.sdkImpl.
Trigger contentended or manually call adsLoader.contentComplete() when a user forcibly switches to a new video.
If an ad is currently playing and the user wants to skip, consider stopping the ad before switching content.
You’ve already discovered the most important fix—calling contentComplete()—so you’re on the right track. If you can replace your direct call to controller.sdkImpl.adsLoader with the plugin’s more official adsLoader, that’s a bit cleaner. But if your plugin version simply doesn’t expose it, then, yes, you have to do what you have to do!

**Comment by DharmikUnagar** [comment]
@jg-proboxtv 

1. Why contentComplete() is Essential
The videojs-ima plugin (and indeed the IMA SDK) needs to be told when your main content is finished playing. If you skip that notification (i.e. don’t call contentComplete()), the SDK’s internal ad state machine will not clean itself up, and future ads can get into a weird or broken state.

Whether you call


`player.ima.adsLoader.contentComplete();`
or use the more “internal” route


`player.ima.controller.sdkImpl.adsLoader.contentComplete();
`the important thing is simply to invoke it at the correct time—i.e., once your content is really done.

Why some folks end up calling controller.sdkImpl.adsLoader:
In older versions of videojs-ima, the plugin didn’t expose adsLoader publicly. So people resorted to referencing the internal controller object. Fortunately, in more recent versions you often can (and should) call the public player.ima.adsLoader.

2. Prefer the Public API (If Available)
Before dropping down to controller.sdkImpl, try:

```
if (this.player.ima && this.player.ima.adsLoader) {
  this.player.ima.adsLoader.contentComplete();
}
```
A lot of folks find that it “crashes” or is undefined if the plugin version is older or if the plugin hasn’t yet fully loaded its internal objects. If that’s happening, you can wait until the ad plugin has fully initialized, or ensure you’re only calling it once you know the plugin is ready. That might involve a small tweak to your plugin lifecycle (for example, waiting for an event like adsready).

If that never works in your environment, then yes—controller.sdkImpl.adsLoader is the fallback. But keep in mind, it’s not guaranteed to be stable across plugin upgrades.

3. Use the contentended Event (Where Possible)
The [videojs-ima documentation](https://github.com/googleads/videojs-ima#quick-start) suggests that if your player fires a contentended event on its own (many do), the plugin will automatically call contentComplete() behind the scenes. In that scenario, you can do something like:

```
onEnd = () => {
  // Let video.js know content is done. The plugin will see this and handle the ad teardown.
  this.player.trigger('contentended');

  // Move on to next item:
  this.nextIndex++;
  this.play();
};
```
Sometimes, though, video.js players or custom-coded logic do not automatically fire contentended. If that’s your situation, manually triggering the event (as above) is often the simplest way to achieve the same result.

4. Handling the “Click During Ad” Issue
You mentioned you still see odd behavior if a user clicks on a playlist item while an ad is still running. Typically, you want to ensure that:

If the ad is currently playing, you either (a) let the ad finish, or (b) gracefully stop the current ad playback. Stopping an ad break mid-stream often means you should call adsManager.stop() or something similar via the plugin.
If you do plan to skip the ad (by letting the user click a new video in the playlist), you’ll want to call contentComplete() for the old item so the plugin cleans up. Then, proceed with the new content (which triggers contentchanged, etc.).
Pseudocode-ish example:
```
onPlaylistItemClick = (event, index) => {
  // If an ad is running, gracefully stop it.
  if (this.player.ima && this.player.ima.adsActive) {
    try {
      // The plugin has some methods for this. 
      // For some older versions, you'll do something like:
      // this.player.ima.controller.adsManager.stop();
      // or a combination of methods. 
      this.player.ima.adsManager.stop();
    } catch (err) {
      console.warn('Could not stop ads', err);
    }
  }

  // Let IMA know content was completed, so it resets properly.
  if (this.player.ima && this.player.ima.adsLoader) {
    this.player.ima.adsLoader.contentComplete();
  }

  // Now we actually move to the new video...
  this.nextIndex = index;
  this.play();
};
```
Which approach you take (stopping vs. waiting) depends on the user experience you want. But the main idea is that if you want to forcibly skip or end the current content’s ad, you should also tell IMA that that content is effectively finished so the plugin state resets. This prevents the “double-scrub bar” scenario where the ad’s progress bar might remain stuck.

5. Summary / Best Practices
Always call contentComplete() on your main content finishing (and also if you’re forcibly moving on).
Use the plugin’s public adsLoader if possible, rather than controller.sdkImpl.
Trigger contentended or manually call adsLoader.contentComplete() when a user forcibly switches to a new video.
If an ad is currently playing and the user wants to skip, consider stopping the ad before switching content.
You’ve already discovered the most important fix—calling contentComplete()—so you’re on the right track. If you can replace your direct call to controller.sdkImpl.adsLoader with the plugin’s more official adsLoader, that’s a bit cleaner. But if your plugin version simply doesn’t expose it, then, yes, you have to do what you have to do!
